src/libsyntax/parse/parser.rs

```diff
@@ -3782,8 +3782,13 @@ impl<'a> Parser<'a> {
     }
 
     fn is_catch_expr(&mut self) -> bool {
-        self.token.is_keyword(keywords::Catch) &&
-        self.look_ahead(1, |t| *t == token::OpenDelim(token::Brace))
+        if self.restrictions.contains(Restrictions::RESTRICTION_NO_STRUCT_LITERAL) {
```

Comment:
Not that it matters, but this would be cleaner as `!self.restrictions.contains(...) && self.token.is_keyword(keywords::Catch) && ...`